File: sdk/@launchdarkly/observability-android/lib/src/main/kotlin/com/launchdarkly/observability/network/ErrorRecoverability.kt
package com.launchdarkly.observability.network

import java.net.HttpURLConnection

/**
 * Implemented by exceptions that carry the recoverability verdict of the request that failed, so the
 * classification is made once — where the status code and the GraphQL `extensions` are still
 * available — rather than re-derived from an error message.
 */
interface RecoverableFailure {
    val isRecoverable: Boolean
}

/**
 * Classifies request failures as recoverable (retrying may succeed) or unrecoverable (permanent for
 * this launch). Lives outside Session Replay so every caller of the public graph can share one verdict
 * instead of re-deriving it from status codes.
 */
object ErrorRecoverability {
    private const val HTTP_TOO_MANY_REQUESTS = 429

    /**
     * Tests whether an HTTP error status represents a condition that might resolve on its own if we
     * retry.
     *
     * @param statusCode the HTTP status
     * @return `true` if retrying makes sense; `false` if it should be considered a permanent failure
     */
    fun isHttpErrorRecoverable(statusCode: Int): Boolean {
        if (statusCode !in 400..499) return true

        return when (statusCode) {
            HttpURLConnection.HTTP_BAD_REQUEST, // bad request
            HttpURLConnection.HTTP_CLIENT_TIMEOUT, // request timeout
            HTTP_TOO_MANY_REQUESTS -> true // too many requests
            else -> false // all other 4xx errors are unrecoverable
        }
    }

    /**
     * Classifies any failure thrown by [GraphQLClient]. Errors of unknown origin are treated as
     * recoverable: retrying costs a backed-off request, while a wrong permanent verdict silently disables
     * a feature for the rest of the launch.
     */
    fun isErrorRecoverable(error: Throwable): Boolean = when (error) {
        is GraphQLClientException -> isRecoverable(error)
        is RecoverableFailure -> error.isRecoverable
        else -> true
    }

    private fun isRecoverable(error: GraphQLClientException): Boolean = when (error) {
        // A rejected request can still carry a GraphQL envelope, and an explicit `retryable` there is more
        // specific than the status code, so it wins.
        is GraphQLClientException.HttpStatus ->
            retryableFlag(error.errors) ?: isHttpErrorRecoverable(error.statusCode)

        // The public graph reports rejections as `200` + `errors`, so these are classified like a generic
        // 4xx: unrecoverable unless the server marks an error retryable.
        is GraphQLClientException.GraphQLErrors -> retryableFlag(error.errors) ?: false

        // Timeouts, connectivity loss and malformed responses carry no permanent signal.
        is GraphQLClientException.Transport,
        is GraphQLClientException.Decoding,
        is GraphQLClientException.MissingData -> true
    }

    /**
     * The server's retry verdict for a set of GraphQL errors, or `null` when none of them states one.
     */
    private fun retryableFlag(errors: List<GraphQLError>?): Boolean? {
        if (errors.isNullOrEmpty()) return null
        if (errors.any { it.extensions?.retryable == false }) return false
        if (errors.any { it.extensions?.retryable == true }) return true

        return null
    }
}

File: sdk/highlight-run/src/client/utils/error-recoverability.ts
import { ClientError } from 'graphql-request'
import { PublicGraphError } from '../graph/generated/schemas'

type GraphQLErrors = NonNullable<ClientError['response']['errors']>

// Public graph errors that are permanent whatever the transport reports.
const UNRECOVERABLE_ERRORS: string[] = [
	PublicGraphError.BillingQuotaExceeded.toString(),
]

/**
 * Tests whether an HTTP error status represents a condition that might resolve on its own if we
 * retry. 4xx statuses are permanent apart from the three that describe a transient condition;
 * anything else, including 5xx, is worth another attempt.
 */
export const isHttpErrorRecoverable = (statusCode: number): boolean => {
	if (statusCode < 400 || statusCode >= 500) {
		return true
	}

	switch (statusCode) {
		case 400: // bad request
		case 408: // request timeout
		case 429: // too many requests
			return true
		default:
			return false // all other 4xx errors are unrecoverable
	}
}

/**
 * Classifies a failed public graph request as recoverable (retrying may succeed) or unrecoverable
 * (permanent for this page load). Errors of unknown origin count as recoverable: retrying costs a
 * backed-off request, while a wrong permanent verdict silently disables recording.
 */
export const isErrorRecoverable = (error: unknown): boolean => {
	if (!(error instanceof ClientError)) {
		// Offline, DNS and aborted-fetch failures reject with the raw error and carry no
		// permanent signal.
		return true
	}

	const errors = error.response.errors
	if (errors?.some((e) => UNRECOVERABLE_ERRORS.includes(e.message))) {
		return false
	}

	// An explicit `retryable` is more specific than the status code, so it wins.
	const retryable = retryableFlag(errors)
	if (retryable !== undefined) {
		return retryable
	}

	if (error.response.status >= 400) {
		return isHttpErrorRecoverable(error.response.status)
	}

	// The public graph also reports a rejected request as `200` + `errors`. Classify those like a
	// generic 4xx: permanent unless the server marked an error retryable.
	return !errors?.length
}

/** The server's retry verdict for a set of GraphQL errors, or `undefined` when none states one. */
const retryableFlag = (
	errors: GraphQLErrors | undefined,
): boolean | undefined => {
	if (errors?.some((e) => e.extensions?.retryable === false)) {
		return false
	}
	if (errors?.some((e) => e.extensions?.retryable === true)) {
		return true
	}
	return undefined
}
